src/predictive-text-input/predictive-text-input.tsx

```diff
@@ -0,0 +1,222 @@
+    className,
+    "data-testid": testId,
+    selectedOption,
+    minimumCharacters = 3,
+    fetchOptions,
+    placeholder = "Enter here...",
+    readOnly = false,
+    disabled = false,
+    error,
+    valueExtractor,
+    listExtractor,
+    displayValueExtractor,
+    onSelectOption,
+}: PredictiveTextInputProps<T, V>) => {
+    // =============================================================================
+    // CONST, STATE
+    // =============================================================================
+    const inputValue = selectedOption && displayValueExtractor(selectedOption);
+    const [input, setInput] = useState<string>(inputValue);
+    const [searchedInput, setSearchedInput] = useState<string>(inputValue);
+    const [options, setOptions] = useState<T[]>([]);
+    const [isLoading, setIsLoading] = useState<boolean>(true);
+    const [isError, setIsError] = useState<boolean>();
+    const [isOptionSelected, setIsOptionSelected] = useState<boolean>(
+        !!selectedOption
+    );
+    const [prevOptionSelected, setPrevOptionSelected] =
+        useState<T>(selectedOption);
+    const fetchOptionsRef = useRef(fetchOptions);
+
+    // =============================================================================
+    // DEBOUNCE FUNCTIONS
+    // =============================================================================
+
+    const handleFetchOptions = async (input: string) => {
+        setIsError(false);
+        setIsLoading(true);
+        try {
+            const fetchedOptions = await fetchOptionsRef.current(input);
+            setSearchedInput(input);
+            setOptions(fetchedOptions);
+            setIsLoading(false);
+        } catch {
+            setIsError(true);
+        }
+    };
+
+    const fetchOptionsDebounced = useCallback(
+        debounce((input) => handleFetchOptions(input), 500, {
+            leading: false,
+            trailing: true,
+        }),
+        []
+    );
+
+    // =============================================================================
+    // EFFECTS
+    // =============================================================================
+
+    useEffect(() => {
+        fetchOptionsRef.current = fetchOptions;
+    }, [fetchOptions]);
+
+    useEffect(() => {
+        /***
+         * Should not fetch options for the following the scenario :
+         * 1. When input < minimumCharacters
+         * 2. When user selected option from dropdown becomes input
+         * 3. Initial selectedOption passed in as input
+         */
+        if (
+            input &&
+            input.length >= minimumCharacters &&
+            input !== searchedInput
+        ) {
+            fetchOptionsDebounced(input);
+        } else {
+            fetchOptionsDebounced.cancel();
+        }
+
+        if (input === "") {
+            setPrevOptionSelected(undefined);
+        }
+
+        if (selectedOption && input !== displayValueExtractor(selectedOption)) {
```

Comment:
if selected option has changed, should input be updated to reflect the new value?